src/lib/FreeC/Backend/Agda/Converter/Type.hs

```diff
@@ -0,0 +1,87 @@
+--   Type Theory".
+
+module FreeC.Backend.Agda.Converter.Type
+  ( convertType
+  , convertFunctionType
+  , convertConstructorType
+  , renameAgdaTypeVar
+  )
+where
+
+import           Control.Monad                  ( mapM )
+
+import qualified FreeC.Backend.Agda.Syntax     as Agda
+import           FreeC.Backend.Agda.Converter.Free
+                                                ( free
+                                                , applyFreeArgs
+                                                )
+import           FreeC.Environment.LookupOrFail ( lookupAgdaIdentOrFail )
+import           FreeC.Environment.Renamer      ( renameAndDefineAgdaTypeVar )
+import qualified FreeC.IR.Syntax               as IR
+import           FreeC.Monad.Converter          ( Converter )
+
+-- | Lifts a single type.
+convertType :: IR.Type -> Converter Agda.Expr
+convertType = dagger
+
+-- | Functions not defined in terms lambdas (@f x = …@ not @f = \x -> …@) aren't
+--   evaluated until fully applied, i.e. cannot be bottom. Therefore interleaving
+--   monadic layers isn't needed.
+--
+--   > (τ₁ → τ₂ → … → τₙ)' = τ₁' → Free τ₂' → … → Free τₙ'
+convertFunctionType :: [IR.Type] -> Converter Agda.Expr
+convertFunctionType types = foldr1 Agda.fun <$> mapM dagger types
+
+-- | Constructors aren't evaluated until fully applied. Furthermore the return
+--   type of a constructor for a data type D has to be D and therefore cannot be
+--   lifted.
+--
+--   > (τ₁ → τ₂ → … → τₙ → D)' = τ₁' → Free τ₂' → … → τₙ' → D
+convertConstructorType :: [IR.Type] -> Converter Agda.Expr
+convertConstructorType types =
+  -- We can use the @star@ translation for the data type, because only the name
+  -- and the application of type and @Size@ variables have to be translated.
+  Agda.fun <$> convertFunctionType (init types) <*> star (last types)
+
+-- | Utility function for introducing a new Agda type variable to the current
+--   scope.
+renameAgdaTypeVar :: IR.TypeVarDecl -> Converter Agda.Name
+renameAgdaTypeVar (IR.TypeVarDecl srcSpan name) =
+  Agda.unqualify <$> renameAndDefineAgdaTypeVar srcSpan name
+
+-------------------------------------------------------------------------------
+-- Translations                                                              --
+-------------------------------------------------------------------------------
+
+-- | The dagger translation as describer by Abel et al.
+--   Normally the dagger translations separates mono and poly types, but the
+--   polytype cases are not applicable for our IR.
+--   Note: If polytypes are added in the future (e.g. for type classes) add
+--   their type translation here.
+--
+--   > -- not applicable (polytypes)
+--   > (∀α:κ.σ)’ = (α : κ’) → σ’ -- RankNTypes translate to pi types
+--   > (σ ↦ τ)’ = σ’ → τ’        -- functions handling polytypes e.g.
+--   >                           -- result from type class dictionary translation
+--   > -- all other cases (monotypes):
+--   > τ’ = m τ
```

Comment:
Strictly speaking, there are some polytypes already in the form of `IR.TypeSchema`, they are just handled somewhere else (i.e., in type signatures). It's true that higher rank polymorphism is not supported and would have to be implemented here, but there are also no plans to add support for the `RankNTypes` language extension at the moment. Thus, I wouldn't mention monotypes and polytypes at all. They just confuse the reader.

```suggestion
--   > τ’ = Free τ*
```

Instead, the comment should specify what is done by the function. For example, in this case the function `converts a type from IR to Agda by lifting it into the @Free@ monad`. If you want you can add that `this corresponds to the dagger translation for monotypes as describer by Abel et al.`.